### PR TITLE
[14.0][MIG] l10n_eu_product_adr

### PR DESCRIPTION
**Pull request by StefanRijnhart**
```
14.0.1.0.0 (2021-05-04)
~~~~~~~~~~~~~~~~~~~~~~~
* [BREAKING] Gather dangerous goods attributes on a new model adr.goods,
  and replace all product attributes with a Many2one on this model.
  Remove all other unused or unclear attributes.
* [ADD] XML data for the dangerous goods table from the 2019 edition,
  including translations in 3 languages.
```

### CODE REVIEW

**Comment by StefanRijnhart** [comment]
@i-vyshnevska I have observed some points to improve this module; https://github.com/OCA/community-data-files/pull/109/files#diff-d967fda5116ea9b63e06dd49e2557fbfc802b8b6476598adcc50acd9d8b890b5R49-R73. Do you agree, or are there good reasons for some of the current design choices? What part of the specifications does `great.class` refer to?

**Comment by i-vyshnevska** [comment]
@StefanRijnhart your comments looks reasonable, what I can say would be good to have one more POV as work was based on just one customer requirements, also related report appeared to be more specific for the country

**Comment by simahawk** [comment]
@StefanRijnhart is this ready now?

**Comment by StefanRijnhart** [comment]
@simahawk thank you for your interest. I've now pushed the latest version as accepted by our customer. I believe it is more correct as it allows for multiple entries in the dangerous goods table with the same UN number, distinguished by different packing instructions or limited quantities. It does mean however that the already very bare migration from the previous version was invalidated entirely so I removed it.

File: l10n_eu_product_adr/models/adr_class.py
Comment (i-vyshnevska) [inline l10n_eu_product_adr/models/adr_class.py:8]: you can add a migration here as we renamed clasess

File: l10n_eu_product_adr/models/adr_label.py
Comment (i-vyshnevska) [inline l10n_eu_product_adr/models/adr_label.py]:
```suggestion
            if label.class_ids:
```
I see class_id not class_ids and `ondelete='restrict'` should solve this issue

**Review by i-vyshnevska — COMMENTED** [review]
in new version we not distinguish good and waste

File: l10n_eu_product_adr/models/adr_class.py
Comment (StefanRijnhart) [reply l10n_eu_product_adr/models/adr_class.py:8]: Adding *only* the migration of this class would not be wrong, but not very useful by itself I'm afraid.

File: l10n_eu_product_adr/models/adr_label.py
Comment (StefanRijnhart) [reply l10n_eu_product_adr/models/adr_label.py]: Thanks, I missed this at some point during the refactoring. I've changed this to check the linked goods, and added tests for this method and others.

**Comment by StefanRijnhart** [comment]
> in new version we not distinguish good and waste

@i-vyshnevska sorry, I don't know what you mean by this comment. Can you clarify?

File: l10n_eu_product_adr/scripts/import_adr_multilang_xlsx.py
Comment (mmequignon) [inline l10n_eu_product_adr/scripts/import_adr_multilang_xlsx.py:12]: Please wrap this import with a `try`, `except`
https://github.com/OCA/odoo-community.org/blob/master/website/Contribution/CONTRIBUTING.rst#32imports

File: l10n_eu_product_adr/models/adr_goods.py
Comment (mmequignon) [inline l10n_eu_product_adr/models/adr_goods.py:78]: We had to implement that as well. I don't recall why, but the our customer has some cases where the unnumber isn't a 4 digits long string. Here's where I implemented the solution we agreed on https://github.com/OCA/delivery-carrier/compare/0278821af40dc8fef2fc2518f53038a37bb07deb..0758137ff35a9a0e3170b641d566749de802df00

File: l10n_eu_product_adr/models/product_template.py
Comment (mmequignon) [inline l10n_eu_product_adr/models/product_template.py]: Since the dangerous attributes can vary from a variant to another, we moved those fields from `product.template` to `product.variant`.
see this PR https://github.com/OCA/community-data-files/pull/112

Could you please do the same?

**Review by mmequignon — APPROVED** [review]
functional tests ok, even if it'll break a lot of module which depends on this one ^^'
However, @i-vyshnevska knows better about this.

**Comment by i-vyshnevska** [comment]
> > in new version we not distinguish good and waste
> 
> @i-vyshnevska sorry, I don't know what you mean by this comment. Can you clarify?

previously we had two classes to distinct dangerous "good" and "waste" (products which producing dangerous staff after they used) they treated differently in reports etc

**Comment by Highcooley** [comment]
> > > in new version we not distinguish good and waste
> > 
> > 
> > @i-vyshnevska sorry, I don't know what you mean by this comment. Can you clarify?
> 
> previously we had two classes to distinct dangerous "good" and "waste" (products which producing dangerous staff after they used) they treated differently in reports etc

since our customer does not work with dangerous waste at the moment, I think we can ignore this.

File: l10n_eu_product_adr/models/adr_goods.py
Comment (mmequignon) [inline l10n_eu_product_adr/models/adr_goods.py]: This one should be different
```suggestion
    limited_quantity = fields.Float()
    limited_quantity_uom = fields.Many2one("uom.uom")
```

File: l10n_eu_product_adr/models/product_template.py
Comment (mmequignon) [inline l10n_eu_product_adr/models/product_template.py:32]:
```suggestion
    )
    limited_quantity = fields.Boolean("Limited Qty", compute="_compute_limited_quantity", store=True)
```
With the compute method like so
```python
@api.onchange("adr_goods_id", "adr_goods_id.limited_quantity_uom", "adr_goods_id.limited_quantity", "uom_id", )
def _compute_limited_quantity(self):
    for record in self:
        compare product's uom_qty with adr.goods uom_qty:
        product.limited_qty = adr.good uom_qty < product uom_qty
```

We should also ensure than the `adr.label` uom has the same uom family as the product one.

**Review by mmequignon — COMMENTED** [review]
@StefanRijnhart 
We discussed internally about this, and it does not provide a way to tell that a product is valid or not, regarding the limited_qty.
I provided you a possible solution in the comments.
Thanks!

File: l10n_eu_product_adr/scripts/import_adr_multilang_xlsx.py
Comment (StefanRijnhart) [reply l10n_eu_product_adr/scripts/import_adr_multilang_xlsx.py:12]: Ah, but this is not necessary in recent versions of Odoo. The linked page (now) mentions this as well:
```
This rule doesn't apply neither to Odoo >= v12, as an unmet dependency in an uninstalled module doesn't block the service thanks to this commit:
```

File: l10n_eu_product_adr/models/adr_goods.py
Comment (StefanRijnhart) [reply l10n_eu_product_adr/models/adr_goods.py:78]: It would be interesting to hear when such a number can appear, because no such number appears in any of the specs I've seen!

File: l10n_eu_product_adr/models/product_template.py
Comment (StefanRijnhart) [reply l10n_eu_product_adr/models/product_template.py]: Alright, I will take care of that.

File: l10n_eu_product_adr/models/adr_goods.py
Comment (StefanRijnhart) [reply l10n_eu_product_adr/models/adr_goods.py]: OK. I will not commit your suggestion directly because of the implications, but I will have a look at this.

File: l10n_eu_product_adr/models/product_template.py
Comment (StefanRijnhart) [reply l10n_eu_product_adr/models/product_template.py:32]: I'm sorry, but my customer is a bulk trader. They trade in pallets, or gallons, and the individual subpackaging is encoded in the product name rather than the UoM of the product. So what suits your requirements is not generally valid.

**Comment by mmequignon** [comment]
@StefanRijnhart 
Hello Stefan, I see that you put this PR back in progress.
Could you please list what you will do, and give us a raw estimate of how many time it will take?
Thank you!

**Comment by StefanRijnhart** [comment]
@mmequignon I tried to indicate in my comments what I was going to refactor:

* https://github.com/OCA/community-data-files/pull/109#discussion_r745969757
* https://github.com/OCA/community-data-files/pull/109#discussion_r745970606

I hope to finish it this week on Monday or Tuesday.

File: l10n_eu_product_adr/models/adr_goods.py
Comment (StefanRijnhart) [reply l10n_eu_product_adr/models/adr_goods.py]: See new commit "[IMP] l10n_eu_product_adr: distinct uom for limited quantity" in this PR for this.

File: l10n_eu_product_adr/models/product_template.py
Comment (StefanRijnhart) [reply l10n_eu_product_adr/models/product_template.py]: See new commit `[RFR] l10n_eu_product_adr: allow different adr goods per variant`

**Review by mmequignon — APPROVED** [review]

**Comment by mmequignon** [comment]
We checked this with the customer, and there's a few missing fields:
- hazard_ind https://github.com/OCA/community-data-files/blob/13.0/l10n_eu_product_adr/models/product_product.py#L50
- packaging_group https://github.com/OCA/community-data-files/blob/13.0/l10n_eu_product_adr/models/product_product.py#L52-L54
- envir_hazardous https://github.com/OCA/community-data-files/blob/13.0/l10n_eu_product_adr/models/product_product.py#L31-L33 (for this one, I'll ask how to determine its value)

[adr-tool](https://adr-tool.com/1339/un-1993)

**Comment by simahawk** [comment]
@i-vyshnevska can you update your review?
@StefanRijnhart do you think we can merge this?

**Comment by StefanRijnhart** [comment]
@simahawk that would be good. The Camptocamp team can then propose follow-up PRs for improvements as they see fit.

File: l10n_eu_product_adr/__manifest__.py
Comment (i-vyshnevska) [inline l10n_eu_product_adr/__manifest__.py:7]:
```suggestion
    "version": "14.0.1.0.0",
```

File: l10n_eu_product_adr/models/adr_goods.py
Comment (i-vyshnevska) [inline l10n_eu_product_adr/models/adr_goods.py:81]: do you need to override name_search if un_number is already in name_get?

File: l10n_eu_product_adr/models/product_template.py
Comment (i-vyshnevska) [inline l10n_eu_product_adr/models/product_template.py:16]: I would rather forbid having under one template dangerous and not dangerous product, it is hard to imagine such situation

File: l10n_eu_product_adr/models/product_template.py
Comment (i-vyshnevska) [inline l10n_eu_product_adr/models/product_template.py:27]: same hare if products have different classes what class will have a template?

File: l10n_eu_product_adr/models/stock_picking.py
Comment (i-vyshnevska) [inline l10n_eu_product_adr/models/stock_picking.py]:
```suggestion
    @api.depends("move_lines.product_id", "move_lines.product_uom_qty)
```

File: l10n_eu_product_adr/scripts/import_adr_multilang_xlsx.py
Comment (i-vyshnevska) [inline l10n_eu_product_adr/scripts/import_adr_multilang_xlsx.py:309]: this is very customer specific

File: l10n_eu_product_adr/__manifest__.py
Comment (StefanRijnhart) [reply l10n_eu_product_adr/__manifest__.py:7]: I'm sorry, but I changed the datamodel on your request and added the related migration for my customer and for anyone else who might have been running this code so I'm not taking this suggestion.

File: l10n_eu_product_adr/models/product_template.py
Comment (StefanRijnhart) [reply l10n_eu_product_adr/models/product_template.py:16]: That seemed to me to be a necessary consequence of @mmequignon's request to allow different settings per product variant. If it is not allowed, it is not possible to extend the number of variants by adding product attribute options on the template because the actual dangerous good is not available to the Odoo mechanism that expands the number of variants.

File: l10n_eu_product_adr/models/product_template.py
Comment (StefanRijnhart) [reply l10n_eu_product_adr/models/product_template.py:27]: It will be assigned the class of the first variant. This is necessary to keep compatibility for my customer who does not have different classes per variant.

File: l10n_eu_product_adr/scripts/import_adr_multilang_xlsx.py
Comment (StefanRijnhart) [reply l10n_eu_product_adr/scripts/import_adr_multilang_xlsx.py:309]: This is not customer specific, it is specific to the sheet that contains the ADR data. It is made available by a Belgian organization, and as a consequence it contains a limited amount of Dutch terminology.

File: l10n_eu_product_adr/models/adr_goods.py
Comment (StefanRijnhart) [reply l10n_eu_product_adr/models/adr_goods.py:81]: The un_number is included in the representation of the record returned by `name_get`, but it is not included in the record's name so I believe the answer is yes.

File: l10n_eu_product_adr/__manifest__.py
Comment (sebalix) [reply l10n_eu_product_adr/__manifest__.py:7]: Let's keep this version then.

**Comment by sebalix** [comment]
I think we cannot merge this as-is because:

- some fields are still missing, see the following comment: https://github.com/OCA/community-data-files/pull/109#issuecomment-975698882 (if we do this in a new PR that would mean that users migrating from 13.0 to 14.0 with this version will get the missing columns dropped from the DB until we add the missing fields in a new version, is it ok to do this?)
- we should rebase & merge https://github.com/StefanRijnhart/community-data-files/pull/1 to improve the migration script

**Comment by StefanRijnhart** [comment]
I can add hazard_ind and packaging_group, as these are also present in the sheet that I import the classes from. I was waiting if more information about environmentally hazardous was coming, but I think that maybe it is obsolete
image
(screenshot from https://adrbook.com/en/2017/ADR/2.2.9.1.9)

File: l10n_eu_product_adr/security/ir.model.access.csv
Comment (TDu) [inline l10n_eu_product_adr/security/ir.model.access.csv]: Could the read access be set on the `base.group_user` instead ?
I see access error for non inventory user opening product pages.
Same question for all read access in this file.

Comment (StefanRijnhart) [reply l10n_eu_product_adr/security/ir.model.access.csv]: Thanks for your comment, I'll check how to prevent the access error.

Comment (sebalix) [reply l10n_eu_product_adr/security/ir.model.access.csv]: Indeed, we are facing some access errors when encoding an SO because of this new data model `adr.label`. @StefanRijnhart did you check it or you prefer we open a PR on your branch?

Comment (StefanRijnhart) [reply l10n_eu_product_adr/security/ir.model.access.csv]: @sebalix I have a fix in mind for the product form, but if you run into problems creating an SO then it would be interesting to have a traceback.

Comment (StefanRijnhart) [reply l10n_eu_product_adr/security/ir.model.access.csv]: Fix pushed. Rather than extending the access rights, I've added groups on the product views to prevent the accdess error. I also apply `compute_sudo` to the computation of stock move's adr points, I hope this helps when validating a sales order.

**Comment by rousseldenis** [comment]
@StefanRijnhart What's the status of this ?

**Comment by bosd** [comment]
Can the generation of a new Runboat be triggered?

**Comment by marylla** [comment]
> Can the generation of a new Runboat be triggered?

@rousseldenis @StefanRijnhart Do you know how to do it? On details it just says {"detail":"Not Found"}...

**Comment by rousseldenis** [comment]
@StefanRijnhart Could your rebase ?

**Comment by StefanRijnhart** [comment]
I've rebased. As for the status, I'm considering to let this one slip because as expected this does not align well with the team that provided the previous implementation, and I think the communication has not been easy for either side. Also, I'm not working for the project that required this anymore and I'm a bit of a hippie so I prefer to work with bicycles rather than with dangerous goods anyway :grin:

**Comment by marylla** [comment]
@bosd The runboat is now working, so you can test it. :) I would too but I don't know enough about this topic to test it properly.

**Review by bosd — APPROVED** [review]
LGTM!

**Comment by marylla** [comment]
@jgrandguillaume Can we please merge this one?

**Comment by bosd** [comment]
Maybe @OCA/community-data-files-maintainers can help

**Comment by simahawk** [comment]
FYI On our side we are waiting for @mmequignon to get back to this for a final review.
We have many other modules depending (and blocked) on this and we need to check how they are impacted after the last changes.
To name some:
* https://github.com/OCA/wms/pull/334
* https://github.com/OCA/delivery-carrier/pull/441
* https://github.com/OCA/stock-logistics-workflow/pull/937

**Review by mmequignon — APPROVED** [review]

**Pull request merged**
